File: src/main/jira-integration/jira.js
import { ipcMain } from "electron";
import { requireArgParams } from "../infrastructure/handler-helper.js";
import axios from "axios";
import { IPC_EVENTS  } from '@common/ipc-events';
let username = "";
let apiToken = "";
let address = "";
let secureStorage;
let settings;
let conn;
let window;

function init(sett, sec, win) {
  secureStorage = sec;
  settings = sett;
  window = win;

  // 注册 IPC 事件监听器
  ipcMain.on(IPC_EVENTS.JIRA.REPO_CHANGED, requireArgParams(initJira, ["id"]));
  ipcMain.on(IPC_EVENTS.JIRA.GET_ISSUE, requireArgParams(getIssue, ["key"]));
  ipcMain.on(
    IPC_EVENTS.JIRA.UPDATE_ISSUE,
    requireArgParams(updateIssue, ["key", "data"])
  );
  ipcMain.on(IPC_EVENTS.JIRA.ADD_COMMENT, requireArgParams(addComment, ["key", "body"]));
  ipcMain.on(
    IPC_EVENTS.JIRA.GET_ASSIGNABLE_USERS,
    requireArgParams(findAssignableUsers, ["key"])
  );
  ipcMain.on(
    IPC_EVENTS.JIRA.ASSIGN_ISSUE,
    requireArgParams(assignIssue, ["key", "name"])
  );
  ipcMain.on(IPC_EVENTS.JIRA.ADD_SUBTASK, requireArgParams(addSubtask, ["key", "name"]));
  ipcMain.on(IPC_EVENTS.JIRA.SEARCH_ISSUES, requireArgParams(searchIssues, ["jql"]));
}

function initJira(event, arg) {
  if (settings.get("jira-enabled")) {
    username = settings.get("jira-username");
    address = settings.get("jira-address");
    secureStorage.getPass(`jira-token@${arg.id}`).then((pass) => {
      apiToken = pass;
      if (username && apiToken && address) {
        conn = axios.create({
          baseURL: `https://${address}/rest/api/2`,
          timeout: 10 * 1000,
          headers: {
            Authorization: `Basic ${Buffer.from(
              username + ":" + apiToken
            ).toString("base64")}`,
          },
        });
        // setup response inteceptors
        conn.interceptors.response.use(
          function (response) {
            return response;
          },
          function (error) {
            if (error.response) {
              if (
                error.response.headers[IPC_EVENTS.JIRA.LOGIN_REASON] ===
                "AUTHENTICATION_DENIED"
              ) {
                window.webContents.send(IPC_EVENTS.JIRA.CAPTCHA_REQUIRED, {});
                return Promise.resolve(error);
              } else if (error.response.status === 404) {
                window.webContents.send(IPC_EVENTS.JIRA.NOT_FOUND, {});
                return Promise.reject(error);
              } else if (error.request.path.indexOf("search") !== -1) {
                // swallow search errors
                return Promise.resolve("QUERY_ISSUE");
              } else {
                window.webContents.send(IPC_EVENTS.JIRA.OPERATION_FAILED, {});
              }
            } else if (error.code === "ECONNABORTED") {
              window.webContents.send(IPC_EVENTS.JIRA.TIMEOUT, { error: error });
            } else {
              window.webContents.send(IPC_EVENTS.JIRA.ERROR, { error: error });
            }
            return Promise.reject(error);
          }
        );
        getResolution();
        getIssueTypes();
      } else {
        conn = null;
      }
    });
  } else {
    conn = null;
  }
}

function getIssue(event, arg) {
  if (conn) {
    return getJiraIssue(arg.key).then((result) => {
      checkStoryFields(result);
      event.sender.send(IPC_EVENTS.JIRA.ISSUE_RETRIEVED, { issue: result.data });
    });
  }
}

function addComment(event, arg) {
  if (conn) {
    let data = {
      body: arg.body,
    };
    return conn.post(`/issue/${arg.key}/comment`, data).then((result) => {
      return getIssue(event, arg);
    });
  }
}

function getJiraIssue(key) {
  if (conn) {
    return conn
      .get(
        `/issue/${key}?expand=renderedFields,names,transitions,transitions.fields,editmeta`
      )
      .then((result) => {
        result.data.fields.description = result.data.renderedFields.description;
        return result;
      });
  } else {
    return Promise.reject("NO_CONN");
  }
}

function getResolution() {
  if (conn) {
    return conn.get(`/resolution`).then((result) => {
      window.webContents.send(IPC_EVENTS.JIRA.RESOLUTIONS_RETRIEVED, {
        resolutions: result.data,
      });
    });
  }
}

function getIssueTypes() {
  if (conn) {
    return conn.get("/issuetype").then((result) => {
      let subtaskType;
      result.data.forEach((issueType) => {
        if (issueType.subtask) {
          subtaskType = issueType;
        }
      });
      window.webContents.send(IPC_EVENTS.JIRA.ISSUE_TYPES_RETRIEVED, {
        issueTypes: result.data,
        subtaskType: subtaskType,
      });
    });
  }
}

function updateIssue(event, arg) {
  if (conn) {
    let req;
    if (arg.data.transition) {
      req = conn.post(`/issue/${arg.key}/transitions`, arg.data);
    } else {
      req = conn.put(`/issue/${arg.key}`, arg.data);
    }
    return req.then((result) => {
      return getIssue(event, arg);
    });
  }
}

function checkStoryFields(expandedResult) {
  Object.keys(expandedResult.data.names).forEach((k) => {
    // check story points
    if (expandedResult.data.names[k] === "Story Points") {
      expandedResult.data.fields.storyPoints = expandedResult.data.fields[k];
    }
  });
}

function findAssignableUsers(event, arg) {
  if (conn) {
    let url = `/user/assignable/search?issueKey=${arg.key}`;
    if (arg.search) {
      url += "&username=" + arg.search;
    }
    return conn.get(url).then((result) => {
      let resp = { key: arg.key, result: result.data };
      event.sender.send(IPC_EVENTS.JIRA.ASSIGNABLE_USERS_RETRIEVED, { result: resp });
    });
  }
}

function assignIssue(event, arg) {
  if (conn) {
    return conn
      .put(`/issue/${arg.key}/assignee`, { name: arg.name })
      .then((result) => {
        getIssue(event, arg);
      });
  }
}

function addSubtask(event, arg) {
  if (conn) {
    return conn
      .post(`/issue`, {
        fields: {
          project: {
            id: arg.projectId,
          },
          parent: {
            key: arg.key,
          },
          issuetype: {
            id: arg.subtaskId,
          },
          summary: arg.name,
        },
      })
      .then((result) => {
        getIssue(event, arg);
      });
  }
}

function searchIssues(event, arg) {
  if (conn) {
    let url = `/search`;
    let obj = { jql: arg.jql };
    if (arg.fields) {
      obj.fields = arg.fields;
    }
    return conn.post(url, obj).then((resp) => {
      if (resp === "QUERY_ISSUE") {
        event.sender.send(IPC_EVENTS.JIRA.ISSUE_QUERY_RESULT_RETRIEVED, { issues: [] });
      } else {
        event.sender.send(IPC_EVENTS.JIRA.ISSUE_QUERY_RESULT_RETRIEVED, {
          issues: resp.data.issues,
        });
      }
    });
  }
}

export { init };

File: apps/desktop/src-tauri/src/jira.rs
//! JIRA REST API v2 integration.
//!
//! Replicates `src/main/jira-integration/jira.js` using `reqwest`.
//! All functions are self-contained and stateless — the HTTP client is
//! built per-request from credentials stored in the OS keychain.

use base64::Engine;
use reqwest::blocking::Client;
use serde_json::Value;

const TIMEOUT_SECS: u64 = 10;

/// Build a JIRA REST API client with Basic Auth.
/// Returns `None` if required settings are missing.
pub fn build_client(repo_id: &str) -> Option<(Client, String)> {
    let enabled = crate::settings_store::get("jira-enabled")?;
    if enabled != "true" && enabled != "1" {
        return None;
    }

    let username = crate::settings_store::get("jira-username")?;
    let address = crate::settings_store::get("jira-address")?;
    let token_key = format!("jira-token@{repo_id}");
    let api_token = crate::secure::get_password(&token_key)?;

    if username.is_empty() || api_token.is_empty() || address.is_empty() {
        return None;
    }

    let auth = base64::engine::general_purpose::STANDARD
        .encode(format!("{username}:{api_token}"));
    let base_url = format!("https://{address}/rest/api/2");

    let client_with_auth = Client::builder()
        .timeout(std::time::Duration::from_secs(TIMEOUT_SECS))
        .default_headers({
            let mut headers = reqwest::header::HeaderMap::new();
            headers.insert(
                reqwest::header::AUTHORIZATION,
                format!("Basic {auth}").parse().unwrap(),
            );
            headers
        })
        .build()
        .ok()?;

    Some((client_with_auth, base_url))
}

/// GET /issue/{key}?expand=...
pub fn get_issue(client: &Client, base_url: &str, key: &str) -> Result<Value, String> {
    let url = format!(
        "{base_url}/issue/{key}?expand=renderedFields,names,transitions,transitions.fields,editmeta"
    );
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    let mut data: Value = resp.json().map_err(|e| e.to_string())?;

    if let Some(rendered_desc) = data.pointer("/renderedFields/description").cloned() {
        if let Some(fields) = data.get_mut("fields").and_then(|f| f.as_object_mut()) {
            fields.insert("description".to_string(), rendered_desc);
        }
    }

    check_story_fields(&mut data);
    Ok(data)
}

/// POST /issue/{key}/comment
pub fn add_comment(
    client: &Client,
    base_url: &str,
    key: &str,
    body: &str,
) -> Result<Value, String> {
    let url = format!("{base_url}/issue/{key}/comment");
    client
        .post(&url)
        .json(&serde_json::json!({ "body": body }))
        .send()
        .map_err(|e| e.to_string())?;
    get_issue(client, base_url, key)
}

/// PUT /issue/{key} or POST /issue/{key}/transitions
pub fn update_issue(
    client: &Client,
    base_url: &str,
    key: &str,
    data: &Value,
) -> Result<Value, String> {
    let is_transition = data.get("transition").is_some();
    if is_transition {
        let url = format!("{base_url}/issue/{key}/transitions");
        client
            .post(&url)
            .json(data)
            .send()
            .map_err(|e| e.to_string())?;
    } else {
        let url = format!("{base_url}/issue/{key}");
        client
            .put(&url)
            .json(data)
            .send()
            .map_err(|e| e.to_string())?;
    }
    get_issue(client, base_url, key)
}

/// PUT /issue/{key}/assignee
pub fn assign_issue(
    client: &Client,
    base_url: &str,
    key: &str,
    name: &str,
) -> Result<Value, String> {
    let url = format!("{base_url}/issue/{key}/assignee");
    client
        .put(&url)
        .json(&serde_json::json!({ "name": name }))
        .send()
        .map_err(|e| e.to_string())?;
    get_issue(client, base_url, key)
}

/// GET /user/assignable/search?issueKey={key}&username={search}
pub fn find_assignable_users(
    client: &Client,
    base_url: &str,
    key: &str,
    search: Option<&str>,
) -> Result<Value, String> {
    let mut url = format!("{base_url}/user/assignable/search?issueKey={key}");
    if let Some(s) = search {
        url.push_str(&format!("&username={s}"));
    }
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    resp.json().map_err(|e| e.to_string())
}

/// POST /issue (create subtask)
pub fn add_subtask(
    client: &Client,
    base_url: &str,
    key: &str,
    name: &str,
    project_id: &str,
    subtask_id: &str,
) -> Result<Value, String> {
    let url = format!("{base_url}/issue");
    client
        .post(&url)
        .json(&serde_json::json!({
            "fields": {
                "project": { "id": project_id },
                "parent": { "key": key },
                "issuetype": { "id": subtask_id },
                "summary": name,
            }
        }))
        .send()
        .map_err(|e| e.to_string())?;
    get_issue(client, base_url, key)
}

/// POST /search
pub fn search_issues(
    client: &Client,
    base_url: &str,
    jql: &str,
    fields: Option<&[String]>,
) -> Result<Value, String> {
    let url = format!("{base_url}/search");
    let mut body = serde_json::json!({ "jql": jql });
    if let Some(f) = fields {
        body["fields"] = serde_json::json!(f);
    }
    let resp = client.post(&url).json(&body).send().map_err(|e| e.to_string())?;
    let data: Value = resp.json().map_err(|e| e.to_string())?;
    Ok(data.get("issues").cloned().unwrap_or(serde_json::json!([])))
}

/// GET /resolution
pub fn get_resolutions(client: &Client, base_url: &str) -> Result<Value, String> {
    let url = format!("{base_url}/resolution");
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    resp.json().map_err(|e| e.to_string())
}

/// GET /issuetype
pub fn get_issue_types(client: &Client, base_url: &str) -> Result<Value, String> {
    let url = format!("{base_url}/issuetype");
    let resp = client.get(&url).send().map_err(|e| e.to_string())?;
    resp.json().map_err(|e| e.to_string())
}

/// Extract Story Points from custom fields into `fields.storyPoints`.
fn check_story_fields(data: &mut Value) {
    let sp_field_key = if let Some(names) = data.get("names").and_then(|n| n.as_object()) {
        names
            .iter()
            .find(|(_, v)| v.as_str() == Some("Story Points"))
            .map(|(k, _)| k.clone())
    } else {
        None
    };

    if let Some(key) = sp_field_key {
        if let Some(value) = data.pointer(&format!("/fields/{key}")).cloned() {
            if let Some(fields) = data.get_mut("fields").and_then(|f| f.as_object_mut()) {
                fields.insert("storyPoints".to_string(), value);
            }
        }
    }
}
